src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -812,6 +813,75 @@ public function simplePaginate($perPage = null, $columns = ['*'], $pageName = 'p
         ]);
     }
 
+    /**
+     * Paginate the given query into a cursor paginator.
+     *
+     * @param  int|null  $perPage
+     * @param  array  $columns
+     * @param  string  $cursorName
+     * @param  string|null  $cursor
+     * @return \Illuminate\Contracts\Pagination\Paginator
+     * @throws \Exception
+     */
+    public function cursorPaginate($perPage = null, $columns = ['*'], $cursorName = 'cursor', $cursor = null)
+    {
+        $cursor = $cursor ?: CursorPaginator::resolveCurrentCursor($cursorName);
+
+        $perPage = $perPage ?: $this->model->getPerPage();
+
+        $orders = $this->ensureOrderForCursorPagination(! is_null($cursor) && $cursor->isPrev());
+
+        $orderDirection = $orders->first()['direction'] ?? 'asc';
+
+        $comparisonOperator = ($orderDirection === 'asc' ? '>' : '<');
+
+        $parameters = $orders->pluck('column')->toArray();
+
+        if (count($parameters) === 1 && ! is_null($cursor)) {
+            $this->where($column = $parameters[0], $comparisonOperator, $cursor->getParam($column));
+        } elseif (count($parameters) > 1 && ! is_null($cursor)) {
+            $this->whereRowValues($parameters, $comparisonOperator, $cursor->getParams($parameters));
+        }
+
+        $this->take($perPage + 1);
+
+        return $this->cursorPaginator($this->get($columns), $perPage, $cursor, [
+            'path' => Paginator::resolveCurrentPath(),
+            'cursorName' => $cursorName,
+            'parameters' => $parameters,
+        ]);
+    }
+
+    /**
+     * Ensure the proper order by required for cursor pagination.
+     *
+     * @param  bool  $shouldReverse
+     * @return \Illuminate\Support\Collection
+     * @throws \Exception
+     */
+    protected function ensureOrderForCursorPagination($shouldReverse = false)
+    {
+        $orderDirections = collect($this->query->orders)->pluck('direction')->unique();
+
+        if ($orderDirections->count() > 1) {
+            throw new Exception('Only a single order by direction is supported in cursor pagination.');
```

Comment:
A dedicated exception class for this would be nice.

Reply:
Hmm, would be great if someone could suggest a name for the exception class. I suck at naming, most of my accepted PRs include class name changes by Taylor 😄

Reply:
I would probably go for a more general `CursorPaginationException` that is used for different kinds of errors. Another choice would be hyper-specific single purpose classes such as `SingleOrderByDirectionException`.

Reply:
Have included this change, thanks! 👍